File: src/algo/sort/QuickSort.java
package algo.sort;

import java.util.Arrays;
import java.util.List;

class QuickSort {

    static List<Integer> unsortedList = Arrays.asList(9, 6, 5, 0, 8, 2, 7, 1);

    public static void main(String[] args) {
        performQuickSort(0, unsortedList.size() - 1);
        printList(unsortedList);
    }

    static int performPartitioning(int startIndex, int endIndex) {
        int pivot = unsortedList.get(endIndex);
        int i = startIndex - 1;
        for (int j = startIndex; j < endIndex; j++) {
            if (unsortedList.get(j) <= pivot) {
                i += 1;
                performSwap(i, j);
            }
        }
        performSwap(i + 1, endIndex);
        return i + 1;
    }

    static void performSwap(int i, int j) {
        int temp = unsortedList.get(i);
        unsortedList.set(i, unsortedList.get(j));
        unsortedList.set(j, temp);
    }

    static void performQuickSort(int startIndex, int endIndex) {
        int middleIndex;
        if (startIndex < endIndex) {
            middleIndex = performPartitioning(startIndex, endIndex);
            performQuickSort(startIndex, middleIndex - 1);
            performQuickSort(middleIndex + 1, endIndex);
        }
    }

    static void printList(List<Integer> sortedList) {
        for (Integer integer : sortedList) {
            System.out.print(integer + " ");
        }
        System.out.println();
    }
}

File: src/algo/sort/QuickSort.kt
package algo.sort

/**
 * @author SYAM K
 */

fun main() {
    val integerList = mutableListOf(22, 33, 11, 55, 99, 77, 66, 88)
    performQuickSort(integerList = integerList, startIndex = 0, endIndex = integerList.size - 1)
    println(integerList)
}

fun performQuickSort(integerList: MutableList<Int>, startIndex: Int, endIndex: Int) {
    if (startIndex < endIndex) {
        val partitionIndex = performPartition(integerList, startIndex, endIndex)
        performQuickSort(integerList = integerList, startIndex = startIndex, endIndex = partitionIndex - 1)
        performQuickSort(integerList = integerList, startIndex = partitionIndex + 1, endIndex = endIndex)
    }
}

fun performPartition(integerList: MutableList<Int>, startIndex: Int, endIndex: Int): Int {
    val pivot = integerList[endIndex]
    var i = startIndex - 1
    for (j in startIndex until endIndex) {
        if (integerList[j] <= pivot) {
            i += 1
            swap(integerList = integerList, i = i, j = j)
        }
    }
    swap(integerList = integerList, i = i + 1, j = endIndex)
    return i + 1
}

fun swap(integerList: MutableList<Int>, i: Int, j: Int) {
    if (i != j && integerList[i] != integerList[j]) {
        integerList[i] = integerList[i] + integerList[j]
        integerList[j] = integerList[i] - integerList[j]
        integerList[i] = integerList[i] - integerList[j]
    }
}
